test/pynrn/_pyobj_testing.py

```diff
@@ -0,0 +1,175 @@
+
+        def __init__(self, arg):
+            self.bp = self.baseattr("p")
+
+        def p(self):
+            self.bp()
+            return self.x
+
+    a = A1(5)
+    assert a.x == 5.0
+    assert a.p() == 6.0
+    b = A1(4)
+    a.s = "one"
+    b.s = "two"
+    assert a.s == "one"
+    assert b.s == "two"
+    assert neuron.h.A[0].s == "one"
+    assert a.p() == 7.0
+    assert b.p() == 5.0
+    a.a = 2
+    b.a = 3
+    assert a.a == 2
+    assert b.a == 3
+    assert neuron.h.List("A").count() == 2
+    a = 1
+    b = 1
+    assert neuron.h.List("A").count() == 0
+
+
+def test_pyobj_constructor():
+    # Test that __new__ is required when __init__ is overridden
+    with pytest.raises(TypeError):
+
+        class PyObj(neuron.HocBaseObject, hoc_type=neuron.h.List):
+            def __init__(self, first):
+                super().__init__()
+                self.append(first)
+
+    class PyObj(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        def __new__(cls, first):
+            return super().__new__(cls)
+
+        def __init__(self, first):
+            super().__init__()
+            self.append(first)
+
+    p = PyObj(neuron.h.List())
+    assert p.count() == 1
+
+
+def test_pyobj_def():
+    class PyObj(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        def my_method(self, a):
+            return a * 2
+
+    p = PyObj()
+    assert p.my_method(4) == 8
+
+
+def test_pyobj_overloading():
+    class PyObj(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        def append(self, i):
+            self.last_appended = i
+            return self.baseattr("append")(i)
+
+    p = PyObj()
+    p2 = PyObj()
+    assert p.append(p) == 1
+    assert p.count() == 1
+    assert p[0] == p
+
+
+def test_pyobj_inheritance():
+    class PyObj(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        pass
+
+    class MyObj(PyObj):
+        pass
+
+    with pytest.raises(TypeError):
+
+        class MyObj2(PyObj):
+            def __init__(self, arg):
+                pass
+
+    class List(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        def __new__(cls, *args, **kwargs):
+            super().__new__(cls)
+
+    class InitList(List):
+        def __init__(self, *args):
+            super().__init__()
+            for arg in args:
+                self.append(arg)
+
+    l = InitList(neuron.h.List(), neuron.h.List())
+
+
+def test_pyobj_composition():
+    class A(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        pass
+
+    class B(neuron.HocBaseObject, hoc_type=neuron.h.List):
+        pass
+
+    class C(neuron.HocBaseObject, hoc_type=neuron.h.Vector):
+        pass
+
+    with pytest.raises(TypeError):
+        # Composition of different HOC types is impossible.
+        class D(A, C):
+            pass
+
+    class E(A, B):
+        pass
+
+    assert E._hoc_type == neuron.h.List
+
+
+class PickleTest(neuron.HocBaseObject, hoc_type=neuron.h.NetStim):
+    def __new__(cls, *args, **kwargs):
```

Comment:
Just curious. I believe that, at least prior to this, and after a lot of effort, the only pickleable hoc object was hoc.Vector.
Does this mean that just about any hoc object is pickleable when wrapped in a python subclass as long as the python subclass can access the entire data comprising the hoc object.  And just needs a  ```__reduce__``` implementation that returns a pickleable tuple and an ```__init__``` that takes that, (now unpickled) tuple as its arg.

Reply:
Yes, that seems like a fair assessment. Things that have no or only non-HOC constructor arguments should be very easy to pickle, as we just reduce them to their class and constructor arguments. For things with HOC object dependencies it depends on how advanced the default unpickler is, but using whatever `h.same` uses to get object identity we can build a map and dependency tree of HOC objects in the pickle.